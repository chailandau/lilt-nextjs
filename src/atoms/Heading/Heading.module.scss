.heading {
    text-transform: uppercase;
    &.centered {
        text-align: center;
    }
    &.green {
        color: color(neon, green);
    }

    &.blue {
        color: color(neon, blue);
    }

    &.md,
    &.xl {
        @include font-recharge-bold;
        text-shadow: shadow(fuchsia);
    }

    &.xs {
        @include font(20, 28);
        @include tablet {
            @include font(26, 36);
        }
    }

    &.sm {
        @include font(24, 34);
        @include tablet {
            @include font(32, 38);
        }
    }

    &.md {
        @include font(22, 26, 0.4);
        @include tablet {
            @include font(38, 55, 0.75);
        }
    }

    &.lg {
        @include font(28, 36);
        @include tablet {
            @include font(44, 54);
        }
        @include desktop {
            @include font(50, 60);
        }
    }

    &.xl {
        @include font(26, 32, 0.5);
        @include tablet {
            @include font(44, 54, 0.85);
        }
        @include desktop {
            @include font(60, 72, 1.2);
        }
    }
}
