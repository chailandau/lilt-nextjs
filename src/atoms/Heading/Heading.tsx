import classNames from 'classnames';
import { FC } from 'react';

import styles from './Heading.module.scss';

export const headingTags = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p'] as const;

export const headingSizes = ['xs', 'sm', 'md', 'lg', 'xl'] as const;

export const headingColors = ['blue', 'green'] as const;

interface HeadingProps {
    /** Semantic tag heading is wrapped in */
    as?: (typeof headingTags)[number];
    /** Heading size */
    size?: (typeof headingSizes)[number];
    /** Heading color */
    color?: (typeof headingColors)[number];
    /** Heading content */
    children: string;
    /** If true, heading will be centered */
    centered?: boolean;
}

const Heading: FC<HeadingProps> = ({
    as: HeadingTag = 'h2',
    size = 'md',
    color = 'green',
    centered,
    children
}) => {
    const classList = classNames(
        styles['heading'],
        styles[size],
        styles[color],
        centered && styles['centered']
    );

    return <HeadingTag className={classList}>{children}</HeadingTag>;
};

export default Heading;
