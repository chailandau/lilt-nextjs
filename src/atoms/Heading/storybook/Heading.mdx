import { Canvas, Meta, ArgTypes } from '@storybook/blocks';

import * as HeadingStories from './Heading.stories';

<Meta of={HeadingStories} />

# Heading Atom

<Canvas of={HeadingStories.Default} />

<br />

## children

Text is supplied as a child to the `Heading` atom.

## as

Select a semantic heading tag using best practices to keep the page accessible. Read this article from [The A11y Project]('https://www.a11yproject.com/posts/how-to-accessible-heading-structure/#best-practices-summarized') for more information on best practices.

If the `as` prop is not defined, the tag will default to `h2`.

## color

Select a color for the `Heading` atom. If no color is selected, color will default to `green`.

## size

Select a size for the `Heading` atom. If no size is selected, size will default to `md`.

<br />
<br />
<br />

## Props Table

<ArgTypes />
