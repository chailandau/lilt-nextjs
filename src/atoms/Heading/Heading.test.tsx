import { composeStory } from '@storybook/react';
import { render, screen } from '@testing-library/react';

import { headingColors, headingSizes, headingTags } from './Heading';
import Meta, {
    Default,
    ExtraLarge,
    ExtraSmall,
    Large,
    Small
} from './storybook/Heading.stories';

import {
    testAxeViolations,
    testMatchesSnapshot,
    testPropOptions,
    testTags
} from '@/utils/testHelpers';

const getHeadingInfo = (Variant: typeof Default) => ({
    variant: composeStory(Variant, Meta),
    text: Variant.args?.children || '',
    color: Variant.args?.color || ''
});

const headingVariants = [
    getHeadingInfo(Default),
    getHeadingInfo(ExtraSmall),
    getHeadingInfo(Small),
    getHeadingInfo(Large),
    getHeadingInfo(ExtraLarge)
];

describe('Heading', () => {
    const Heading = composeStory(Default, Meta);
    testPropOptions({
        component: <Heading />,
        propName: 'size',
        propOptions: headingSizes,
        htmlTag: 'h2'
    });
});

headingVariants.forEach(({ variant: Heading, text }) => {
    const headingComponent = {
        component: <Heading />
    };
    describe(text, () => {
        it('has defaults', () => {
            render(
                <Heading as={undefined} color={undefined} size={undefined} />
            );
            expect(screen.getByRole('heading')).toHaveClass('green', 'md');
            expect(screen.getByRole('heading', { level: 2 })).toBeDefined;
        });
        testTags({ ...headingComponent, tags: headingTags });
        testPropOptions({
            ...headingComponent,
            propName: 'color',
            propOptions: headingColors,
            htmlTag: 'h2'
        });
        testAxeViolations(headingComponent);
        testMatchesSnapshot(headingComponent);
    });
});
