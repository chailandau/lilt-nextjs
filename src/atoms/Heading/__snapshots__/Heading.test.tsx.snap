// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default heading matches snapshot 1`] = `
<DocumentFragment>
  <h2
    class="heading md green"
  >
    Default heading
  </h2>
</DocumentFragment>
`;

exports[`Extra large heading matches snapshot 1`] = `
<DocumentFragment>
  <h2
    class="heading xl green"
  >
    Extra large heading
  </h2>
</DocumentFragment>
`;

exports[`Extra small heading matches snapshot 1`] = `
<DocumentFragment>
  <h2
    class="heading xs green"
  >
    Extra small heading
  </h2>
</DocumentFragment>
`;

exports[`Large heading matches snapshot 1`] = `
<DocumentFragment>
  <h2
    class="heading lg green"
  >
    Large heading
  </h2>
</DocumentFragment>
`;

exports[`Small heading matches snapshot 1`] = `
<DocumentFragment>
  <h2
    class="heading sm green"
  >
    Small heading
  </h2>
</DocumentFragment>
`;
